### ISSUE TITLE
[Bug] Scheduler fails to schedule a pod due to a race condition

### ISSUE THREAD

**Issue by scheduler-tester** [kubernetes/kubernetes#126700]
### What happened?

A bug in the scheduler increases the time spent on scheduling a pod **from <1 second to 5 minutes**.

We discovered the bug when repeating the steps described in a fixed bug report [#106780](https://github.com/kubernetes/kubernetes/issues/106780). The setting and steps are as follows:

We have 1 node with 32Gi memory, and 3 pods to schedule: 
p1 / request 10Gi memory / low-priority,
p2 / request 25Gi memory / medium-priority,
p3 / request 20Gi memory / high-priority.

We perform the following steps:
1. add `p1` and wait until `p1` is running
2. add `p2` and wait until `p1` is terminating; this is because there are not enough resources to host both pods and `p2` has a higher priority.
**Note that at this point, `p2` is not running yet.**
3. add `p3` and wait until `p3` is running; `p2` is pending because `p2` and `p3` cannot coexist on `node0`.
4. re-add `p1` and wait until `p1` is running; `p1` should be able to run since there are enough resources to host both pods: 10 + 20 < 32

Interestingly, we find that if step 4 happened immediately after step 3, then `p1` fails to get scheduled with the reason `Insufficient Memory` and eventually gets scheduled after 5 minutes. If there is some short period between step 3 and step 4 (say 2 seconds), then `p1` is scheduled properly immediately.

#### What is the root cause?

After code and log inspections (see below), we have found the root cause of this bug: it's caused by the race condition between (A) `p1` is re-added and handled by the scheduler and (B) `p2`'s `nominated_node_name` is cleared. If A happens before B, then the bug occurs.

#### Why does 2 second make the difference?

The reason is that `p2`'s `nominated_node_name` is cleared *when it (a nominated pod) is scheduled again and failed*. But as the `p2`'s first schedule failed (need to wait for preemption to finish), it's put to BackoffQueue and need to wait for a few seconds before it's put back to ActionQueue and scheduled again.

So 
1. If `p1` is added before `p2` is scheduled again (without 2s sleep), it will **fail** as `p2`'s `nomianted_node_name` is not cleared.
2. But if we add the 2s sleep, `p2` is scheduled again and its `nominated_node_name` is cleared before `p1` is added. In this case, `p1` will be **schedulable** to node0.

<details><summary>Some scheduler logs of the bug-free and buggy trace</summary>
The lines starting with # are logged just before scheduling p1 and p2. You can notice that their order is reversed in the two traces.

Bug-free trace (B happens before A):
```markdown
# with sleep(2s)
I0814 14:19:47.939731   32287 eventhandlers.go:149] "Add event for unscheduled pod" pod="default/p3"
I0814 14:19:47.939756   32287 schedule_one.go:83] "About to try and schedule pod" pod="default/p3"
I0814 14:19:47.939768   32287 schedule_one.go:96] "Attempting to schedule pod" pod="default/p3"
I0814 14:19:47.939878   32287 default_binder.go:53] "Attempting to bind pod to node" logger="Bind.DefaultBinder" pod="default/p3" node="node0"
I0814 14:19:47.944902   32287 eventhandlers.go:313] "Delete event for scheduled pod" pod="default/p1"
I0814 14:19:47.948208   32287 scheduling_queue.go:1312] "Pod moved to an internal scheduling queue" pod="default/p2" event="AssignedPodDelete" queue="Backoff" hint=1
I0814 14:19:47.948799   32287 schedule_one.go:314] "Successfully bound pod to node" pod="default/p3" node="node0" evaluatedNodes=1 feasibleNodes=1
I0814 14:19:47.948891   32287 eventhandlers.go:201] "Delete event for unscheduled pod" pod="default/p3"
I0814 14:19:47.948901   32287 eventhandlers.go:231] "Add event for scheduled pod" pod="default/p3"
I0814 14:19:47.956662   32287 eventhandlers.go:268] "Update event for scheduled pod" pod="default/p3"
I0814 14:19:49.047040   32287 schedule_one.go:83] "About to try and schedule pod" pod="default/p2"
# I0814 14:19:49.047238   32287 schedule_one.go:96] "Attempting to schedule pod" pod="default/p2"
I0814 14:19:49.047886   32287 schedule_one.go:1055] "Unable to schedule pod; no fit; waiting" pod="default/p2" err="0/1 nodes are available: 1 Insufficient memory. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod."
I0814 14:19:49.048895   32287 schedule_one.go:1122] "Updating pod condition" pod="default/p2" conditionType="PodScheduled" conditionStatus="False" conditionReason="Unschedulable"
I0814 14:19:49.064048   32287 eventhandlers.go:174] "Update event for unscheduled pod" pod="default/p2"
I0814 14:19:49.978428   32287 eventhandlers.go:149] "Add event for unscheduled pod" pod="default/p1"
I0814 14:19:49.978562   32287 schedule_one.go:83] "About to try and schedule pod" pod="default/p1"
# I0814 14:19:49.978580   32287 schedule_one.go:96] "Attempting to schedule pod" pod="default/p1"
I0814 14:19:49.978879   32287 default_binder.go:53] "Attempting to bind pod to node" logger="Bind.DefaultBinder" pod="default/p1" node="node0"
I0814 14:19:49.985438   32287 eventhandlers.go:201] "Delete event for unscheduled pod" pod="default/p1"
I0814 14:19:49.985476   32287 eventhandlers.go:231] "Add event for scheduled pod" pod="default/p1"
I0814 14:19:49.985747   32287 schedule_one.go:314] "Successfully bound pod to node" pod="default/p1" node="node0" evaluatedNodes=1 feasibleNodes=1
```

Buggy trace (A happens before B):
```markdown
# without sleep(2s)
I0814 14:17:47.151445   31615 eventhandlers.go:149] "Add event for unscheduled pod" pod="default/p3"
I0814 14:17:47.151515   31615 eventhandlers.go:174] "Update event for unscheduled pod" pod="default/p2"
I0814 14:17:47.151802   31615 schedule_one.go:83] "About to try and schedule pod" pod="default/p3"
I0814 14:17:47.151812   31615 schedule_one.go:96] "Attempting to schedule pod" pod="default/p3"
I0814 14:17:47.151926   31615 default_binder.go:53] "Attempting to bind pod to node" logger="Bind.DefaultBinder" pod="default/p3" node="node0"
I0814 14:17:47.156395   31615 eventhandlers.go:313] "Delete event for scheduled pod" pod="default/p1"
I0814 14:17:47.156444   31615 scheduling_queue.go:1312] "Pod moved to an internal scheduling queue" pod="default/p2" event="AssignedPodDelete" queue="Backoff" hint=1
I0814 14:17:47.160464   31615 eventhandlers.go:201] "Delete event for unscheduled pod" pod="default/p3"
I0814 14:17:47.160479   31615 eventhandlers.go:231] "Add event for scheduled pod" pod="default/p3"
I0814 14:17:47.160514   31615 schedule_one.go:314] "Successfully bound pod to node" pod="default/p3" node="node0" evaluatedNodes=1 feasibleNodes=1
I0814 14:17:47.168539   31615 eventhandlers.go:268] "Update event for scheduled pod" pod="default/p3"
I0814 14:17:47.177608   31615 eventhandlers.go:149] "Add event for unscheduled pod" pod="default/p1"
I0814 14:17:47.177642   31615 schedule_one.go:83] "About to try and schedule pod" pod="default/p1"
# I0814 14:17:47.177650   31615 schedule_one.go:96] "Attempting to schedule pod" pod="default/p1"
I0814 14:17:47.177775   31615 schedule_one.go:1055] "Unable to schedule pod; no fit; waiting" pod="default/p1" err="0/1 nodes are available: 1 Insufficient memory. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod."
I0814 14:17:47.177802   31615 schedule_one.go:1122] "Updating pod condition" pod="default/p1" conditionType="PodScheduled" conditionStatus="False" conditionReason="Unschedulable"
I0814 14:17:47.187472   31615 eventhandlers.go:174] "Update event for unscheduled pod" pod="default/p1"
I0814 14:17:48.138665   31615 schedule_one.go:83] "About to try and schedule pod" pod="default/p2"
# I0814 14:17:48.139950   31615 schedule_one.go:96] "Attempting to schedule pod" pod="default/p2"
I0814 14:17:48.140305   31615 schedule_one.go:1055] "Unable to schedule pod; no fit; waiting" pod="default/p2" err="0/1 nodes are available: 1 Insufficient memory. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod."
I0814 14:17:48.140582   31615 schedule_one.go:1122] "Updating pod condition" pod="default/p2" conditionType="PodScheduled" conditionStatus="False" conditionReason="Unschedulable"
I0814 14:17:48.156537   31615 eventhandlers.go:174] "Update event for unscheduled pod" pod="default/p2"
```
</details>

#### How to fix it?

1. A potential fix is:
Register all plugins that need to be aware of pods to events triggered by updates to a pod's `nominated_node_name`. This way, a failed pod will be retried as soon as any pod's `nominated_node_name` is cleared.

2. If this is too radical and can cause performance issues, another solution is: 
When a pod is bound to a node with nominated pods, either 
a) re-check the previously nominated pods (that nominated to this node), 
b) or clear the `nominated_node_name` of these nominated pods.
In this way, upcoming pods won't fail unreasonably and won't have to wait 5 minutes for the next retry.

### What did you expect to happen?

`p1` should be immediately scheduled.

### How can we reproduce it (as minimally and precisely as possible)?

We're using [kwok](https://kwok.sigs.k8s.io/) to reproduce this issue.

```yaml
# node0.yaml
apiVersion: v1
kind: Node
metadata:
  name: node0
  labels:
    kubernetes.io/hostname: node0
status:
  allocatable:
    cpu: "32"
    memory: "32Gi"
    pods: "110"
  capacity:
    cpu: "32"
    memory: "32Gi"
    pods: "110"
```

```yaml
# p1.yaml
apiVersion: v1
kind: Pod
metadata:
  name: p1
spec:
  containers:
  - name: p1-container
    image: nginx
    resources:
      requests:
        memory: "10Gi"

  priorityClassName: low-priority
```

```yaml
# p2.yaml
apiVersion: v1
kind: Pod
metadata:
  name: p2
spec:
  containers:
  - name: p2-container
    image: nginx
    resources:
      requests:
        memory: "25Gi"
  priorityClassName: medium-priority
```

```yaml
# p3.yaml
apiVersion: v1
kind: Pod
metadata:
  name: p3
spec:
  containers:
  - name: p3-container
    image: nginx
    resources:
      requests:
        memory: "20Gi"
  priorityClassName: high-priority
```

```yaml
# priority_class.yaml
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: high-priority
value: 1000000
globalDefault: false
description: "This priority class should be used for XYZ service pods only."
---
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: medium-priority
value: 500000
globalDefault: false
description: "This priority class should be used for XYZ service pods only."
---
apiVersion: scheduling.k8s.io/v1
kind: PriorityClass
metadata:
  name: low-priority
value: 0
globalDefault: false
description: "This priority class should be used for XYZ service pods only."
```

```python
# To run this reproduction script, please create a new KWOK cluster first.
# kwokctl create cluster --v DEBUG
# Also need to install the kubernetes python client library,
# and place all yaml files in the same folder with the script.
# ---
# To check the scheduler logs:
# kwokctl logs kube-scheduler
# ---
# To reproduce the bug case, please comment out `sleep(2)` between step 5 and step 6.
# To reproduce the normal case, please add `sleep(2)` between step 5 and step 6.
import unittest
import time
import shutil
import logging
from time import sleep
from os import path, makedirs
from logging import getLogger
from kubernetes import config, watch
from kubernetes.client import *
from kubernetes.utils import *
from time import strftime

logger = getLogger(__name__)
config.load_kube_config()
v1 = CoreV1Api()
k8s_cli = ApiClient()


log_dir = path.dirname(__file__)

if __name__ == "__main__":
    log_path = path.join(log_dir, f"issue-106780-reproduction-{strftime('%Y-%m-%d-%H-%M-%S')}")
    shutil.rmtree(log_path, ignore_errors=True)
    makedirs(log_path, exist_ok=True)
    log_file = path.join(log_path, 'reproduce.log')
    logging.basicConfig(level=logging.INFO,
                        filemode='w',
                        filename=log_file,
                        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                        datefmt='%H:%M:%S')
    node0_yaml = path.join(path.dirname(__file__), 'node0.yaml')
    p1_yaml = path.join(path.dirname(__file__), 'p1.yaml')
    p2_yaml = path.join(path.dirname(__file__), 'p2.yaml')
    p3_yaml = path.join(path.dirname(__file__), 'p3.yaml')
    priority_class_yaml = path.join(path.dirname(__file__), 'priority_class.yaml')
    w = watch.Watch()

    # 1. create priority class
    create_from_yaml(k8s_client=k8s_cli, yaml_file=priority_class_yaml)
    logger.info(f"PriorityClass created")

    # 2. add node0
    create_from_yaml(k8s_client=k8s_cli, yaml_file=node0_yaml)
    logger.info(f"Node: node0 created")

    # 3. add p1
    create_from_yaml(k8s_client=k8s_cli, yaml_file=p1_yaml)
    logger.info(f"Pod: p1 added")

    # 3. check p1 is running
    p1_running = False
    for event in w.stream(v1.list_namespaced_pod,
                          field_selector=f'metadata.name=p1',
                          namespace="default",
                          timeout_seconds=int(5)):
        p1 = event['object']
        if event['object'].status.phase == "Running":
            p1_running = True
            logger.info(f"Pod: {p1.metadata.name} scheduled on node {p1.spec.node_name}")
            break
    assert p1_running

    # 4. add p2
    create_from_yaml(k8s_client=k8s_cli, yaml_file=p2_yaml)
    logger.info(f"Pod: p2 added")

    # 4. check p1 is terminating (by p2)
    for event in w.stream(v1.list_namespaced_pod,
                          field_selector=f'metadata.name=p1',
                          namespace="default",
                          timeout_seconds=int(3)):
        logger.info(f"Pod: {event['object'].metadata.name}, event: {event['type']}")
        logger.info(
            f"deletion_timestamp: {event['object'].metadata.deletion_timestamp}, status: {event['object'].status.phase}")
        if event['object'].metadata.deletion_timestamp is not None and event['object'].status.phase in (
        'Pending', 'Running'):
            logger.info(f"Pod: {p1.metadata.name} is terminating")
            break

    # 5. add p3 and wait for p3 running (p3 will terminate p2 / let p2 pending)
    create_from_yaml(k8s_client=k8s_cli, yaml_file=p3_yaml)
    p3_running = False
    for event in w.stream(v1.list_namespaced_pod,
                          field_selector=f'metadata.name=p3',
                          namespace="default",
                          timeout_seconds=int(5)):
        p3 = event['object']
        if p3.status.phase == "Running":
            logger.info(f"Pod: {p3.metadata.name} scheduled on node p3")
            p3_running = True
            break
    assert p3_running

    # sleep(2)  # -> without this on 1.30.2 will cause pod scheduled failed at the first try

    # 6. reapply and check p1 schedulable
    create_from_yaml(k8s_client=k8s_cli, yaml_file=p1_yaml)
    tic = time.time()
    logger.info(f"Pod: p1 re-applied")

    # 7. watch pod1 events and wait it to be scheduled
    p1_running = False
    for event in w.stream(v1.list_namespaced_pod,
                          field_selector=f'metadata.name=p1',
                          namespace="default",
                          timeout_seconds=int(600)):
        p1 = event['object']
        if p1.status.phase == "Running":
            p1_running = True
            logger.info(f"Pod: {p1.metadata.name} scheduled on node {p1.spec.node_name}")
            break

    if not p1_running:
        logger.error(f"Pod: {p1.metadata.name} is not scheduled in 600s")
        assert False
    else:
        logger.info(f"Pod: {p1.metadata.name}, " + f"time between add and scheduled: {time.time() - tic}")


```

### Anything else we need to know?

Tested on 1.30.2

/sig bugs scheduling

### Kubernetes version

<details>

Tested on 1.30.2

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by chengjoey** [comment]
hi @scheduler-tester ,This is the result I got from running a python script. I can't seem to reproduce it.:
```
09:51:18 - root - INFO - Pod: p1 added
09:51:18 - root - INFO - Pod: p1 scheduled on node node0
09:51:18 - root - INFO - Pod: p2 added
09:51:18 - root - INFO - Pod: p1, event: ADDED
09:51:18 - root - INFO - deletion_timestamp: None, status: Running
09:51:18 - root - INFO - Pod: p1, event: MODIFIED
09:51:18 - root - INFO - deletion_timestamp: 2024-08-16 01:51:48+00:00, status: Running
09:51:18 - root - INFO - Pod: p1 is terminating
09:51:18 - root - INFO - Pod: p3 scheduled on node p3
09:51:21 - root - INFO - Pod: p1 re-applied
09:51:21 - root - INFO - Pod: p1 scheduled on node node0
09:51:21 - root - INFO - Pod: p1, time between add and scheduled: 0.040593862533569336
```

kubectl get pod -o wide after execute:
```
python3 script.py
09:51:18 - root - INFO - Pod: p1 added
09:51:18 - root - INFO - Pod: p1 scheduled on node node0
09:51:18 - root - INFO - Pod: p2 added
09:51:18 - root - INFO - Pod: p1, event: ADDED
09:51:18 - root - INFO - deletion_timestamp: None, status: Running
09:51:18 - root - INFO - Pod: p1, event: MODIFIED
09:51:18 - root - INFO - deletion_timestamp: 2024-08-16 01:51:48+00:00, status: Running
09:51:18 - root - INFO - Pod: p1 is terminating
09:51:18 - root - INFO - Pod: p3 scheduled on node p3
09:51:21 - root - INFO - Pod: p1 re-applied
09:51:21 - root - INFO - Pod: p1 scheduled on node node0
09:51:21 - root - INFO - Pod: p1, time between add and scheduled: 0.040593862533569336
```

**Comment by scheduler-tester** [comment]
@chengjoey  Hi, thanks for your swift reply! 
I'm sorry for my oversight: this bug only occurs without the 2-second sleep (as mentioned in the description above). Therefore, the `sleep(2)` between step 5 and step 6 in the python script should be commented out if you want to reproduce the bug (I've updated the script above to reflect this change).

Below is the log of running the script without `sleep(2)`:
```
10:10:10 - __main__ - INFO - Pod: p1 added
10:10:11 - __main__ - INFO - Pod: p1 scheduled on node node0
10:10:11 - __main__ - INFO - Pod: p2 added
10:10:11 - __main__ - INFO - Pod: p1, event: ADDED
10:10:11 - __main__ - INFO - deletion_timestamp: None, status: Running
10:10:11 - __main__ - INFO - Pod: p1, event: MODIFIED
10:10:11 - __main__ - INFO - deletion_timestamp: None, status: Pending
10:10:11 - __main__ - INFO - Pod: p1, event: MODIFIED
10:10:11 - __main__ - INFO - deletion_timestamp: 2024-08-16 02:10:41+00:00, status: Pending
10:10:11 - __main__ - INFO - Pod: p1 is terminating
10:10:11 - __main__ - INFO - Pod: p3 scheduled on node p3
10:10:11 - __main__ - INFO - Pod: p1 re-applied
10:15:18 - __main__ - INFO - Pod: p1 scheduled on node node0
10:15:18 - __main__ - INFO - Pod: p1, time between add and scheduled: 306.9531891345978
```

`kubectl get pod -o wide`: 1 min after execute
```
NAME   READY   STATUS    RESTARTS   AGE   IP         NODE     NOMINATED NODE   READINESS GATES
p1     0/1     Pending   0          59s   <none>     <none>   <none>           <none>
p2     0/1     Pending   0          59s   <none>     <none>   <none>           <none>
p3     1/1     Running   0          59s   10.0.0.3   node0    <none>           <none>

```
`kubectl get pod -o wide`: 5 minutes after execute
```
NAME   READY   STATUS    RESTARTS   AGE     IP         NODE     NOMINATED NODE   READINESS GATES
p1     1/1     Running   0          5m30s   10.0.0.4   node0    <none>           <none>
p2     0/1     Pending   0          5m30s   <none>     <none>   <none>           <none>
p3     1/1     Running   0          5m30s   10.0.0.3   node0    <none>           <none>
```

**Comment by chengjoey** [comment]
@scheduler-tester thanks, I reproduced it

**Comment by chengjoey** [comment]
hi @scheduler-tester , I found that it is not only because of `nominated_node_name`, but also related to priority. I think it cannot be simply fixed according to 2 in how to fix. Although p1 and p2 are both nominated, if the priority of p2 is not higher than p1, p1 can still be scheduled quickly. If I fix it according to `2`, I think it will destroy some existing logic

https://github.com/kubernetes/kubernetes/blob/c999f9d828009f0a535237aca74ce2e5f6a8088c/pkg/scheduler/framework/runtime/framework.go#L1035-L1045

If I comment out `nodeInfoOut.AddPodInfo(pi)`, p1 can be scheduled to node0 immediately

```
23:51:35 - root - INFO - Pod: p1 added
23:51:35 - root - INFO - Pod: p1 scheduled on node node0
23:51:35 - root - INFO - Pod: p2 added
23:51:35 - root - INFO - Pod: p1, event: ADDED
23:51:35 - root - INFO - deletion_timestamp: 2024-08-16 15:52:05+00:00, status: Running
23:51:35 - root - INFO - Pod: p1 is terminating
23:51:35 - root - INFO - Pod: p3 scheduled on node p3
23:51:35 - root - INFO - Pod: p1 re-applied
23:51:35 - root - INFO - Pod: p1 scheduled on node node0
23:51:35 - root - INFO - Pod: p1, time between add and scheduled: 0.014630317687988281
```

**Comment by chengjoey** [comment]
@scheduler-tester ,I also found an interesting phenomenon. Change p1 in step `4` to another pod `p4`. p4 and p1 need the same resource `10Gi`, but p4's priority is the same as p2's, both are `medium-priority`. p4 also needs to wait 5 minutes to be scheduled to node0.

p4:
```
# p4.yaml
apiVersion: v1
kind: Pod
metadata:
  name: p4
spec:
  containers:
  - name: p4-container
    image: nginx
    resources:
      requests:
        memory: "10Gi"
  // Note that it is medium-priority like p2
  priorityClassName: medium-priority
```

I think pods of the same priority should not assume that resources are already occupied by the nominated node, because they are all of the same priority. This is unfair to the later pods of the same priority, and the resources required by the later pods may be sufficient, but they are assumed to be occupied by the previous pods.

https://github.com/kubernetes/kubernetes/blob/c999f9d828009f0a535237aca74ce2e5f6a8088c/pkg/scheduler/framework/runtime/framework.go#L1035-L1036

**Comment by scheduler-tester** [comment]
Hi @chengjoey, thanks for your help!
>  If I fix it according to 2, I think it will destroy some existing logic

You're right that if we clean all the nominated pods some logic will be broken.
So maybe we can consider:
1) The most direct way to fix the root cause (where the pod's `nominated_node_name` is not cleaned when itself can no longer be bound to the nominated node) is: When a later pod was bound to a node with nominated pods, check each nominated pod individually. If a pod cannot be bound on the node after the current pod is bound, its `nominated_node_name` should be cleaned.
2) Alternatively, we can register the relevent plugins to a new `UpdatePodNominatedNodeName` event. I noticed there are existing events like `UpdatePodLabel` and `UpdatePodScaleDown`, so adding a new event for this fine-grained update might be helpful. However, I'm doubting that `nominated_node_name` is a field managed by the scheduler, not the user, so maybe there're better solution to do this (trigger the pod is retried as soon some `nominated_node_name` is cleared).

> If I comment out `nodeInfoOut.AddPodInfo(pi)`, p1 can be scheduled to node0 immediately

That's true, but I think this should be kept because according to [the comment](https://github.com/kubernetes/kubernetes/blob/c999f9d828009f0a535237aca74ce2e5f6a8088c/pkg/scheduler/framework/runtime/framework.go#L977C2-L977C82), the scheduler need two pass filter (one pass with nominated pods, one pass without) to make sure the decision is compatible with the nominated pods no matter they are scheduled to the nominated node or not. 

---

> I think pods of the same priority should not assume that resources are already occupied by the nominated node, because they are all of the same priority. 

In my opinion, this logic is sound. P1 not only arrives earlier but is also the one preempting other pods. Therefore, it seems reasonable to reserve some resources for it.

> This is unfair to the later pods of the same priority, and the resources required by the later pods may be sufficient, but they are assumed to be occupied by the previous pods.

I think this behavior is intentional (though I'm not sure), so removing it might disrupt existing logic. Maybe we should ask for member's idea on this.
Alternatively, this issue could be addressed by clearing the nominated_node_name of the previous pod, possibly via the fixes mentioned above.

**Comment by sanposhiho** [comment]
It's an interesting corner case.
So, the scheduler should regard the update event of Pods that clears the nominated node info as the Pod deletion event, but only effective for lower priority Pods. 

So, in this issue's example, when `p2`'s nominatedNodeName is cleared, we should regard the update event to `p2` as Pod/Delete, and move `p1` back to the activeQ. (The resource fit plugin, which should be registered as a unsched plugin of `p2` registers Pod/Delete event) 
And, let's say the scheduler also has `p4` that has the highest priority. `p4` shouldn't be requeued to activeQ. That's what I meant by "only effective for lower priority Pods".

> could this affect the discussions in https://github.com/kubernetes/kubernetes/issues/126858?

My current take is that it's a requeueing (event handling) problem, not a preemption problem, and hence it's not much related to that enhancement.

**Comment by dom4ha** [comment]
I think the issue deserves an update. I added triggering the update some time ago, but struggled to write tests that can simulate it.

The tests are there now, but I discovered that with QHints enabled, the the events are ignored, since they don't treat removal of nominated node name as a delete (or basically a change that can release resources). The fix in the fit plugin is obvious, but I need to go though other plugins and check whether and how they should be modified.
